File: diffusion.py
#!/usr/bin/python
import numpy as np
import math
#Establish the variables needed 
#DS is the size variable 
dS = 20
rsize = dS+2
arraySize = np.power(rsize,3)
print(arraySize)
room = np.zeros(arraySize).reshape(rsize,rsize,rsize)
mask = np.zeros(arraySize).reshape(rsize,rsize,rsize)
#Populate the mask
for i in xrange(1,rsize-1):
	for j in xrange(1,rsize-1):
		for k in xrange(1,rsize-1):
			mask[i,j,k] = 1

#Code for wall
#for i in xrange(rsize/2, rsize-1)
#	for j in xrange(rsize/2, rsize-1)
#		mask[i,j,rsize/2] = 0

#Calculate the necessary variables for the problem
diffusion_coefficent = 0.175
room_dimension = 5.0
speed_of_gas_molecules = 250.0
timestep = (room_dimension/speed_of_gas_molecules) / dS
distance_between_blocks = room_dimension / dS
print(diffusion_coefficent, timestep, distance_between_blocks)
DTerm = diffusion_coefficent * timestep / (distance_between_blocks*distance_between_blocks)
diffused = 0
#Add the gas into the room
room[1,1,1] = 1.0e21
#Zero the following variables for saftey
lpass = 0
time = 0.0
ratio = 0.0
#Run until the equilibration point of .99 is reached across the cubes of the room
while ratio<.99:
	#Three loops one for each dimension
	for i in xrange(1,rsize-1):
		for j in xrange(1,rsize-1):
			for k in xrange(1,rsize-1):
				#Do the calculations for all of the six possible side of the cube that gas could diffuse from
				change = (room[i,j,k] - room[i+1,j,k])*DTerm*mask[i+1,j,k]
				room[i,j,k]-= change
				room[i+1,j,k]+=change	
				
				change = (room[i,j,k] - room[i-1,j,k])*DTerm*mask[i-1,j,k]
				room[i,j,k]-=change
				room[i-1,j,k]+=change	
				
				change = (room[i,j,k] - room[i,j+1,k])*DTerm*mask[i,j+1,k]
				room[i,j,k]-=change
				room[i,j+1,k]+=change	
			
				change = (room[i,j,k] - room[i,j-1,k])*DTerm*mask[i,j-1,k]
				room[i,j,k]-=change
				room[i,j-1,k]+=change	
				
				change = (room[i,j,k] - room[i,j,k+1])*DTerm*mask[i,j,k+1]
				room[i,j,k]-=change
				room[i,j,k+1]+=change	
					
				change = (room[i,j,k] - room[i,j,k-1])*DTerm*mask[i,j,k-1]
				room[i,j,k]-=change
				room[i,j,k-1]+=change	
	#Step the time variable and give min and max a value
	time = time + timestep
	maxval = room[1,1,1]
	minval = room[1,1,1]
	#Find the min and max for the whole matrix
	for i in xrange(1,rsize-1):
		for j in xrange(1,rsize-1):
			for k in xrange(1,rsize-1):
				maxval = max(room[i,j,k], maxval)
				minval = min(room[i,j,k], minval)
	#Calculate the ratio
	ratio = minval/maxval

#	print ratio, time 
#Print the time taken to equilibrate
print "Box equilibrated in "
print time
print " seconds of simulated time."

File: diffusion.f90
program diffusion
!Initialize all parameters and variables that are necessary
real room_dimension, speed_of_gas_molecules, diffusion_coefficent, timestep, distance_between_blocks, DTerm
integer i,j,k,l,m,n,z
!Parameter for number of cubes is ds
integer,parameter ::  ds= 50
integer, parameter :: rsize = ds+2 
integer, parameter :: hsize = rsize/2
real lpass, time, ratio 
real*4 change, mval, mival
real*4, Dimension(rsize,rsize,rsize) :: room
integer, Dimension(rsize,rsize,rsize) :: mask
logical, Dimension(rsize,rsize,rsize) :: lmask
!Complete the calculations required for the diffusing of the gas and proper time
!stepping
diffusion_coefficent = 0.175
room_dimension = 5.0
speed_of_gas_molecules = 250.0
timestep = (room_dimension/speed_of_gas_molecules)/ds
distance_between_blocks = room_dimension / ds
        DTerm = diffusion_coefficent * timestep /&
(distance_between_blocks*distance_between_blocks)
!Zero the mask, room, and logical mask arrays
        do i=1, rsize
        do j=1, rsize
        do k=1, rsize
                room(k,j,i) = 0.0
                mask(k,j,i) = 0
                lmask(k,j,i) = .false.
        enddo
        enddo
        enddo

!        print *, room
!        print *, mask
!Correctly populate the mask defining the outside and inside
        do i=2, rsize-1
        do j=2, rsize-1
        do k=2, rsize-1
                mask(k,j,i) = 1
                lmask(k,j,i) = .true.
        enddo
        enddo
        enddo

!Code to make the wall in the logical and normal mask
!        do i=rsize/2, rsize-1
!        do j=rsize/2, rsize-1
!                lmask(j,i,hsize) = .false.
!                mask(j,i,hsize) = 0
!        enddo
!        enddo  

!Place the gas into the room        
        room(2,2,2) = 1.0e21
!Zero the time ratio and lpass variables for saftey
        lpass = 0
        time = 0.0
        ratio = 0.0
!Run until the gas has properly diffused
do while( ratio< .99)
        !Three loops one for each dimension
        do i=2, rsize-1
        do j=2, rsize-1
        do k=2, rsize-1
                !Do the calculation for diffusion for each of the six cubes that
                !may or may not be in contact with the selected cube.
                !Secondarilay reverse the use of the loops from inner to outer
                !for effiency.
                change = (room(k,j,i)-room(k+1,j,i))*DTerm*mask(k+1,j,i)
                room(k,j,i) = room(k,j,i)-change
                room(k+1,j,i) = room(k+1,j,i)+change        
                
                change = (room(k,j,i)-room(k-1,j,i))*DTerm*mask(k-1,j,i)
                room(k,j,i) = room(k,j,i)-change
                room(k-1,j,i) = room(k-1,j,i)+change        
                
                change = (room(k,j,i)-room(k,j+1,i))*DTerm*mask(k,j+1,i)
                room(k,j,i) = room(k,j,i)-change
                room(k,j+1,i) = room(k,j+1,i)+change        
                
                change = (room(k,j,i)-room(k,j-1,i))*DTerm*mask(k,j-1,i)
                room(k,j,i) = room(k,j,i)-change
                room(k,j-1,i) = room(k,j-1,i)+change        
                
                change = (room(k,j,i)-room(k,j,i+1))*DTerm*mask(k,j,i+1)
                room(k,j,i) = room(k,j,i)-change
                room(k,j,i+1) = room(k,j,i+1)+change        
                
                change = (room(k,j,i)-room(k,j,i-1))*DTerm*mask(k,j,i-1)
                room(k,j,i) = room(k,j,i)-change
                room(k,j,i-1) = room(k,j,i-1)+change        
        enddo
        enddo
        enddo 
        !Increment the time
        time = time + timestep
        !Give max and min a value
        mval = room(2,2,2)
        mival = room(2,2,2)
        !Find the max and min of the matrix using the built in function and
        !utilising the logical mask to make such an operation possible
        mval = MAXVAL(room, lmask)
        mival = MINVAL(room, lmask)
        !Find the current ratio of diffusion
        ratio = mival/mval

        !print *, ratio, time
        enddo
        !State the amount of time it took to equilibrate the room
        print *, "Box equilibrated in "
        print *, time
        print *, " seconds of Simulated Time."

        end
